File: example/src/main/java/com/matnsolutions/matlive/lib/screens/audio_room_screen.dart
import 'package:flutter/material.dart';
import 'package:matlive_client_sdk_flutter/mat_live.dart';
import '../widgets/audio_room_layout.dart';

class AudioRoomScreen extends StatefulWidget {
  final String roomId;
  final String token;

  const AudioRoomScreen({
    super.key,
    required this.roomId,
    required this.token,
  });

  @override
  State<AudioRoomScreen> createState() => _AudioRoomScreenState();
}

class _AudioRoomScreenState extends State<AudioRoomScreen> {
  final _matLiveRoomManger = MatLiveRoomManger.instance;
  bool loading = true;

  final TextEditingController controller = TextEditingController();

  @override
  void initState() {
    super.initState();
    _initializeRoom();
  }

  Future<void> _initializeRoom() async {
    try {
      await MatLiveJoinRoomManger.instance.init(
        onInvitedToMic: (seatIndex) {},
        onSendGift: (data) {},
      );
      await MatLiveJoinRoomManger.instance.connect(
        roomId: widget.roomId,
        token: widget.token,
        name: 'Ibrahim Nashatat',
        avatar:
            'https://img-cdn.pixlr.com/image-generator/history/65bb506dcb310754719cf81f/ede935de-1138-4f66-8ed7-44bd16efc709/medium.webp',
        userId: '10',
        metadata: '{userRome:"admin"}',
      );

      final seatService = MatLiveRoomManger.instance.seatService;
      // Initialize seat layout
      seatService!.initWithConfig(
        MatLiveAudioRoomLayoutConfig(
          rowSpacing: 16,
          rowConfigs: [
            MatLiveAudioRoomLayoutRowConfig(count: 4, seatSpacing: 12),
            MatLiveAudioRoomLayoutRowConfig(count: 4, seatSpacing: 12),
          ],
        ),
      );
    } catch (e) {
      if (!mounted) return;
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Failed to connect: $e')),
      );
    }
    loading = false;
    setState(() {});

    /// _matLiveRoomManger.inviteRequests;
  }

  @override
  void dispose() {
    MatLiveJoinRoomManger.instance.close();
    super.dispose();
  }

  void _sendGift(String gift) {
    _matLiveRoomManger.sendGift(gift);
    _showSnackBar('sendGift');
  }

  void _requestTakeMic(int seatIndex) {
    _matLiveRoomManger.requestTakeMic(seatIndex);
    _showSnackBar('requestTakeMic');
  }

  void _clearChat() {
    _matLiveRoomManger.clearChat();
    _showSnackBar('clearChat');
  }

  void _inviteUserToTakeMic(String userId, int seatIndex) {
    _matLiveRoomManger.inviteUserToTakeMic(userId, seatIndex);
    _showSnackBar('inviteUserToTakeMic');
  }

  void _handleTakeMic(int index) {
    _matLiveRoomManger.takeSeat(index);
    _showSnackBar('Took mic at seat $index');
  }

  void _handleMuteMic(int index) {
    final seatService = MatLiveRoomManger.instance.seatService;
    final seat = seatService!.seatList.value[index];
    if (!seat.currentUser.value!.isMicOnNotifier.value) {
      _matLiveRoomManger.muteSeat(index);
    } else {
      _matLiveRoomManger.unMuteSeat(index);
    }
    _showSnackBar(!seat.currentUser.value!.isMicOnNotifier.value
        ? 'Unmuted mic at seat $index'
        : 'Muted mic at seat $index');
  }

  void _handleRemoveSpeaker(int index) {
    _matLiveRoomManger.removeUserFromSeat(index);
    _showSnackBar('Removed speaker from seat $index');
  }

  void _handleLeaveMic(int index) {
    _matLiveRoomManger.leaveSeat(index);
    _showSnackBar('Left mic at seat $index');
  }

  void _handleLockMic(int index) {
    _matLiveRoomManger.lockSeat(index);
    _showSnackBar('Seat locked $index');
  }

  void _handleUnLockMic(int index) {
    _matLiveRoomManger.unLockSeat(index);
    _showSnackBar('Seat unlocked $index');
  }

  void _handleSwitchSeat(int toIndex) {
    _matLiveRoomManger.switchSeat(toIndex);
    _showSnackBar('Switched seat to $toIndex');
  }

  void _showSnackBar(String message) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text(message)),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(
          'Room: ${widget.roomId}',
          style: const TextStyle(
            color: Colors.black,
            fontSize: 16.0,
          ),
        ),
      ),
      body: Padding(
        padding: const EdgeInsets.symmetric(vertical: 21.0),
        child: loading
            ? const Center(
                child: SizedBox(
                  width: 24,
                  height: 24,
                  child: CircularProgressIndicator(
                    strokeWidth: 2,
                  ),
                ),
              )
            : Column(
                children: [
                  Padding(
                    padding: const EdgeInsets.only(bottom: 10.0),
                    child: AudioRoomLayout(
                      onTakeMic: _handleTakeMic,
                      onMuteMic: _handleMuteMic,
                      onRemoveSpeaker: _handleRemoveSpeaker,
                      onLeaveMic: _handleLeaveMic,
                      onSwitchSeat: _handleSwitchSeat,
                      onLockMic: _handleLockMic,
                      onUnLockMic: _handleUnLockMic,
                    ),
                  ),
                  Expanded(
                    child: SingleChildScrollView(
                      padding: const EdgeInsets.symmetric(horizontal: 21.0),
                      child: ValueListenableBuilder<List<MatLiveChatMessage>>(
                        valueListenable: MatLiveRoomManger.instance.messages,
                        builder: (context, messages, _) {
                          return Column(
                            crossAxisAlignment: CrossAxisAlignment.start,
                            children: messages.map(
                              (item) {
                                return Padding(
                                  padding: const EdgeInsets.only(bottom: 8.0),
                                  child: Column(
                                    crossAxisAlignment:
                                        CrossAxisAlignment.start,
                                    children: [
                                      Row(
                                        children: [
                                          Image.network(
                                            item.user.avatar,
                                            fit: BoxFit.cover,
                                            height: 25.0,
                                            width: 25.0,
                                            errorBuilder:
                                                (context, error, stackTrace) {
                                              return const Icon(Icons.person,
                                                  size: 22.0);
                                            },
                                          ),
                                          const SizedBox(width: 12.0),
                                          Text(item.user.name),
                                        ],
                                      ),
                                      const SizedBox(height: 4.0),
                                      Text(item.message),
                                    ],
                                  ),
                                );
                              },
                            ).toList(),
                          );
                        },
                      ),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.all(8.0),
                    child: Row(
                      children: [
                        Expanded(
                          child: TextField(
                            controller: controller,
                            decoration: const InputDecoration(
                              hintText: 'Type a message ...',
                              border: OutlineInputBorder(),
                            ),
                          ),
                        ),
                        IconButton(
                          onPressed: () {
                            if (controller.text.isNotEmpty) {
                              MatLiveRoomManger.instance
                                  .sendMessage(controller.text);
                              controller.clear();
                            }
                          },
                          icon: const Icon(Icons.send),
                        )
                      ],
                    ),
                  )
                ],
              ),
      ),
    );
  }
}

File: example/src/main/java/com/matnsolutions/matlive/lib/screens/AudioRoomScreen.kt
package com.matnsolutions.matlive.lib.screens

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Send
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.matnsolutions.matlive.lib.widgets.AudioRoomLayout
import com.matnsolutions.matlive_sdk.audio.define.MatLiveChatMessage
import com.matnsolutions.matlive_sdk.audio.mangers.MatLiveJoinRoomManger
import com.matnsolutions.matlive_sdk.audio.mangers.MatLiveRoomManger
import com.matnsolutions.matlive_sdk.audio.seats.MatLiveAudioRoomLayoutConfig
import com.matnsolutions.matlive_sdk.audio.seats.MatLiveAudioRoomLayoutRowConfig
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AudioRoomScreen(roomId: String, token: String) {
    val matLiveRoomManger = MatLiveRoomManger.instance
    var loading by remember { mutableStateOf(true) }
    val controller = remember { mutableStateOf("") }
    val scope = rememberCoroutineScope()

    val context = androidx.compose.ui.platform.LocalContext.current
    LaunchedEffect(Unit) {
        MatLiveJoinRoomManger.instance.init(
            onInvitedToMic = {},
            onSendGift = {}
        )
        MatLiveJoinRoomManger.instance.connect(
            context = context,
            roomId = roomId,
            token = token,
            name = "Ahmed Attia",
            avatar = "https://img-cdn.pixlr.com/image-generator/history/6565c8dff9ef18d69df3e3a2/fe1887b5-015e-4421-8c6a-1364d2f5b1e9/medium.webp",
            userId = "10",
            metadata = "{userRome:\"admin\"}"
        )

        val seatService = MatLiveRoomManger.instance.seatService
        seatService?.initWithConfig(
            MatLiveAudioRoomLayoutConfig(
                rowSpacing = 16.0,
                rowConfigs = listOf(
                    MatLiveAudioRoomLayoutRowConfig(count = 4, seatSpacing = 12),
                    MatLiveAudioRoomLayoutRowConfig(count = 4, seatSpacing = 12)
                )
            )
        )

        loading = false
    }

    Scaffold(
        topBar = {
            TopAppBar(title = {
                Text(
                    text = "Room: $roomId",
                    style = TextStyle(color = Color.Black, fontSize = 16.sp)
                )
            })
        }
    ) { paddingValues ->
        Box(modifier = Modifier.padding(paddingValues)) {
            if (loading) {
                Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                    CircularProgressIndicator(modifier = Modifier.size(24.dp), strokeWidth = 2.dp)
                }
            } else {
                Column(modifier = Modifier.padding(vertical = 21.dp)) {
                    AudioRoomLayout(
                        onTakeMic = { scope.launch { matLiveRoomManger.takeSeat(it) } },
                        onMuteMic = { scope.launch { matLiveRoomManger.muteSeat(it) } },
                        onRemoveSpeaker = { scope.launch { matLiveRoomManger.removeUserFromSeat(it) } },
                        onLeaveMic = { scope.launch { matLiveRoomManger.leaveSeat(it) } },
                        onSwitchSeat = { scope.launch { matLiveRoomManger.switchSeat(it) } },
                        onLockMic = { scope.launch { matLiveRoomManger.lockSeat(it) } },
                        onUnLockMic = { scope.launch { matLiveRoomManger.unLockSeat(it) } }
                    )

                    val messages by matLiveRoomManger.messages.collectAsState()

                    Box(modifier = Modifier.weight(1f)) {
                        LazyColumn(
                            modifier = Modifier.padding(horizontal = 21.dp),
                        ) {
                            items(messages) { item ->
                                ChatMessageItem(item = item)
                            }
                        }
                    }

                    Row(
                        modifier = Modifier.padding(8.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        TextField(
                            value = controller.value,
                            onValueChange = { controller.value = it },
                            modifier = Modifier.weight(1f),
                            placeholder = { Text("Type a message ...") },
                            colors = TextFieldDefaults.textFieldColors(
                                containerColor = Color.White,
                            ),
                        )
                        IconButton(onClick = {
                            if (controller.value.isNotEmpty()) {
                                scope.launch {
                                    matLiveRoomManger.sendMessage(controller.value)
                                    controller.value = ""
                                }
                            }
                        }) {
                            Icon(Icons.Filled.Send, contentDescription = "Send")
                        }
                    }
                }
            }
        }
    }
}


@Composable
fun ChatMessageItem(item: MatLiveChatMessage) {
    Column(modifier = Modifier.padding(bottom = 8.dp)) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            AsyncImage(
                model = item.user.avatar,
                contentDescription = "User Avatar",
                modifier = Modifier
                    .size(25.dp)
                    .clip(CircleShape),
                contentScale = ContentScale.Crop,
//                error = {
//                    Icon(Icons.Filled.Person, contentDescription = "User Avatar", modifier = Modifier.size(22.dp))
//                }
            )
            Spacer(modifier = Modifier.width(12.dp))
            Text(text = item.user.name)
        }
        Spacer(modifier = Modifier.height(4.dp))
        Text(text = item.message)
    }
}
